transformer_engine/pytorch/attention.py

```diff
@@ -806,18 +867,68 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
     return torch.cat((-x2, x1), dim=-1)
 
 
-def apply_rotary_pos_emb(t: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
+def apply_rotary_pos_emb(
+    t: torch.Tensor,
+    freqs: torch.Tensor,
+    fused: bool = False,
+    transpose_output_memory: bool = False,
+    qkv_format: str = "sbhd",
+    cu_seqlens: Union[torch.Tensor, None] = None,
+) -> torch.Tensor:
     """
-    input tensor t is of shape [seq_length, ..., dim]
-    rotary positional embeding tensor `freqs` is of shape [seq_length, ..., dim]
+    Apply rotary positional embedding tensor to the input tensor.
+
+    Parameters
+    ----------
+    t: torch.Tensor
+        Tensor of shape [s, b, h, d] or [t, h, d].
+    freqs: torch.Tensor
+        Rotary positional embedding tensor of shape [s, 1, 1, d] and dtype 'float'.
+    fused: bool, default = False
+        Whether to use a fused applying RoPE implementation.
+    transpose_output_memory: bool, default = False
+        Whether to transpose the 's' and 'b' dimension of the output's underlying memory format.
+        This is very helpful when you want to get a contiguous tensor after calling
+        `output.transpose(0, 1)`. It's only supported when `fused` = True and
+        `qkv_format` = 'sbhd'.
+    qkv_format: str, default = 'sbhd'.
+        It could be 'sbhd' or 'thd', and 'thd' is only supported when `fused` = True.
+    cu_seqlens: torch.Tensor, default = None.
+        Cumulative sum of sequence lengths in a batch for `t`, with shape [b + 1] and
+        dtype torch.int32. Only valid when `qkv_format` = 'thd'.
     """
+    if fused:
+        assert qkv_format in (
+            "sbhd",
+            "thd",
+        ), f"Only 'sbhd' and 'thd' formats are supported when fused is True, got {qkv_format}."
+        assert not (
+            transpose_output_memory and qkv_format == "thd"
+        ), "transpose_output_memory is only supported with 'sbhd' format."
+        assert (
+            qkv_format == "sbhd" or cu_seqlens is not None
+        ), "cu_seqlens must not be None when qkv_format is 'thd'."
+        return FusedRoPEFunc.apply(
```

Comment:
Can this function take in input batches of different sequence lengths? (like batch1 is `b, 256, h, d` while batch2 is `b, 240, s, d`). 
Something like this: https://github.com/NVIDIA/TransformerEngine/pull/557/files#diff-0af6d715a51b3efcebd6067805b5d17b64d25ef84399e256bade01a602ce4192R861-R873

I guess we could convert the input to `thd` format for such use-case but I'm wondering if it could still work without doing it?

Reply:
Is it a common case? `freqs` is usually calculated by taking a `max_seq_len` argument here. https://github.com/NVIDIA/TransformerEngine/blob/82555b3fd1ddece3a886fac3c7a4aa8d6f8c1955/transformer_engine/pytorch/attention.py#L809
So in most case, they should have the same seq length I think?

Of course I could add the support if needed.

Reply:
Added it.

Reply:
Ah I didn't know the intended use of the `forward` method in every iteration. Thanks for pointing it out!
I still feel that if the `max_seq_len` is globally the max, then we could simply get slice from this tensor for every smaller batch we get, instead of creating it afresh in every iteration.

Reply:
It's supported now.